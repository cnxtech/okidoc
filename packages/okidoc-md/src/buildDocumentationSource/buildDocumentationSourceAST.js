import fs from 'fs';
import path from 'path';
import * as t from '@babel/types';
import traverseSource from '../utils/traverseSource';
import traverseEntries from '../utils/traverseEntries';
import traverseFiles from '../utils/traverseFiles';
import {
  createApiVisitor,
  createApiMethod,
  createApiInterface,
  buildApiClassDeclaration,
} from '../api';

function ensureSourceOptionsValid({ entry, pattern, source }) {
  if (!entry && !pattern && !source) {
    throw new Error(
      `'entry' or/and 'pattern' or 'source' options should be defined`,
    );
  }

  if (source && (entry || pattern)) {
    throw new Error(
      `if 'source' options provided, 'entry' and 'pattern' should be undefined`,
    );
  }
}

function ensureVisitorExists(apiVisitorPath) {
  if (!fs.existsSync(apiVisitorPath)) {
    throw new Error(
      `'visitor' option should be valid js file path. '${apiVisitorPath}' is not exists`,
    );
  }
}

function ensureCreateApiVisitorIsFunction(createCustomApiVisitor) {
  if (typeof createCustomApiVisitor !== 'function') {
    throw new Error(
      `'visitor' file should export function 'createApiVisitor' (exports.createApiVisitor = ...)`,
    );
  }
}

function ensureCustomApiVisitorIsCorrect(
  visitors,
  customApiVisitor,
  visitorPath,
) {
  const visitorsKeys = Object.keys(visitors);
  const customApiVisitorKeys = Object.keys(customApiVisitor);
  const intersection = customApiVisitorKeys.filter(key =>
    visitorsKeys.includes(key),
  );

  if (intersection.length) {
    console.error(
      `default config visitors ${JSON.stringify(
        intersection,
      )} are overridden by config from 'visitor.createApiVisitor' (${visitorPath})`,
    );
  }
}

function traverse({ entry, pattern, source }, visitors) {
  if (source) {
    traverseSource(source, visitors);
    return;
  }

  if (entry) {
    traverseEntries(entry, visitors, { pattern });
    return;
  }

  traverseFiles(pattern, visitors);
}

function buildDocumentationSourceAST({
  entry,
  pattern,
  source,
  tag,
  visitor: visitorPath,
}) {
  const apiInterfaces = [];
  const classApiMethods = [];
  const functions = [];

  ensureSourceOptionsValid({ entry, pattern, source });

  if (!tag && !visitorPath) {
    throw new Error(
      `'tag' or/and 'visitor' options should be defined ('${pattern}')`,
    );
  }

  // NOTE: read about visitors
  // https://github.com/thejameskyle/babel-handbook/blob/master/translations/en/plugin-handbook.md#toc-visitors
  // https://github.com/babel/babel/blob/master/packages/babel-traverse/src/visitors.js
  const visitors = {
    'InterfaceDeclaration|TSInterfaceDeclaration': path => {
      apiInterfaces.push(createApiInterface(path.node));
    },
  };

  if (tag) {
    Object.assign(
      visitors,
      createApiVisitor(tag, path => {
        if (t.isClassMethod(path.node || t.isClassProperty(path.node))) {
          classApiMethods.push(createApiMethod(path.node, tag));
        }

        if (
          t.isVariableDeclaration(path.node) ||
          t.isExportNamedDeclaration(path.node) ||
          t.isFunctionDeclaration(path.node)
        ) {
          functions.push(createApiMethod(path.node, tag));
        }
      }),
    );
  }

  if (visitorPath) {
    visitorPath = path.resolve(visitorPath);

    ensureVisitorExists(visitorPath);

    const {
      createApiVisitor: createCustomApiVisitor,
      createApiMethod: createCustomApiMethod,
    } = require(visitorPath);

    ensureCreateApiVisitorIsFunction(createCustomApiVisitor);

    const customApiVisitor = createCustomApiVisitor(path => {
      if (t.isClassMethod(path.node || t.isClassProperty(path.node))) {
        classApiMethods.push(
          typeof createCustomApiMethod === 'function'
            ? createCustomApiMethod(path.node)
            : createApiMethod(path.node),
        );
      }
    });

    ensureCustomApiVisitorIsCorrect(visitors, customApiVisitor, visitorPath);

    Object.assign(visitors, customApiVisitor);
  }

  traverse({ entry, pattern, source }, visitors);

  const apiClassDeclaration = buildApiClassDeclaration(classApiMethods);

  return {
    type: 'Program',
    body: apiInterfaces.concat(apiClassDeclaration, functions),
  };
}

export default buildDocumentationSourceAST;
