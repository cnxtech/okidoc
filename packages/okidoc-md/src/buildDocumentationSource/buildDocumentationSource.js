import generate from '@babel/generator';

import buildDocumentationSourceAST from './buildDocumentationSourceAST';

function buildDocumentationSource({
  entry,
  source,
  pattern,
  tag,
  visitor,
  interfaces = {},
}) {
  return generate(
    buildDocumentationSourceAST({
      entry,
      source,
      pattern,
      tag,
      visitor,
      interfaces,
    }),
  ).code;
}

export default buildDocumentationSource;
