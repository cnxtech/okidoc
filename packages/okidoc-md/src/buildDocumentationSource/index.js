import buildDocumentationSource from './buildDocumentationSource';

export default buildDocumentationSource;
