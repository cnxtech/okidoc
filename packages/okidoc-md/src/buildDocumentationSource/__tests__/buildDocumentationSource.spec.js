import path from 'path';
import buildDocumentationSource from '../';

describe('buildDocumentationSource', () => {
  describe('for API class', () => {
    describe('with @doc tag on class members', () => {
      it('should extract methods with tag', () => {
        const sourceCode = `
      class MyComponent {
        /**
         * show
         *
         * @doc UI
         * @example
         * component.show()
         */
        show() {
          console.log('show');
        }
      
        /**
         * Disable
         */
        disable() {
            console.log('disable');
        }
      }`;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });

      it('should extract getters and setters with tag', () => {
        const sourceCode = `
      class MyComponent {
        /**
        * node getter
        */
        get node() {}

        /**
         * \`width\` getter
         *
         * @doc UI
         */
        get width() {
          console.log('get width');
        }
        
        /**
         * \`width\` setter
         *
         * @doc UI
         */
        set width(value) {
          console.log('set width', value);
        }
      }`;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });

      it('should ignore private methods with tag', () => {
        const sourceCode = `
      class MyComponent {
        /**
         * init
         *
         * @doc UI
         */
        private init() {
          console.log('init');
        }
      }`;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });

      it('should ignore methods with different tag', () => {
        const sourceCode = `
      class MyComponent {
        /**
         * show
         *
         * @doc UI
         */
        show() {
          console.log('show');
        }
      
        /**
         * disable
         *
         * @doc Destroy
         */
        destroy() {
            console.log('destroy');
        }
      }`;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });
    });

    describe('with @doc tag on class declaration', () => {
      it('should extract class members', () => {
        const sourceCode = `
      /**
      * @doc UI
      */
      class MyComponent {
        /**
        * \`isHidden\` getter
        */
        get isHidden() {}

        /**
         * show
         */
        show() {
          console.log('show');
        }
      
        /**
         * hide
         */
        hide() {
          console.log('hide');
        }
        
        disable() {
          console.log('hide');
        }
      }`;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });

      it('should ignore class constructor and private members', () => {
        const sourceCode = `
      /**
      * @doc UI
      */
      class MyComponent {
        constructor()
      
        /**
        * init
        */
        private init() {}
      
        render() {
          console.log('render');
        }
      }`;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });
    });

    describe('with custom visitor', () => {
      it('should extract class members', () => {
        const sourceCode = `
        function api() {}
        
        class MyComponent {
          @api()
          get isHidden() {}
  
          /**
           * show
           */
          @api()
          show() {
            console.log('show');
          }
        
          /**
           * hide
           */
          @api()
          hide() {
            console.log('hide');
          }
          
          disable() {
            console.log('hide');
          }
        }`;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            visitor: require.resolve('./fixtures/customApiVisitor.js'),
          }),
        ).toMatchSnapshot();
      });
    });
  });

  describe('for functions', () => {
    describe('with @doc tag on function declaration', () => {
      it('should extract functions with tag', () => {
        const sourceCode = `
      /**
       * initComponent
       */
      function initComponent() {
        console.log('initComponent');
      }

      /**
       * showComponent
       *
       * @doc UI
       */
      function showComponent() {
        console.log('showComponent');
      }
      
      /**
       * hideComponent
       *
       * @doc UI
       */
      function hideComponent() {
        console.log('hideComponent');
      }
      `;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });
    });

    describe('with @doc tag on var/let/const declaration', () => {
      it('should extract functions with tag', () => {
        const sourceCode = `
      /**
       * funcToIgnore
       */
      var funcToIgnore = function () {
        console.log('funcToIgnore');
      }

      /**
       * varFunc
       * @doc UI
       */
      var varFunc = function () {
        console.log('varFunc');
      }
      
      /**
       * varArrowFunc
       * @doc UI
       */
      var varArrowFunc = () => {
        console.log('varArrowFunc');
      }

      /**
       * letArrowFunc
       * @doc UI
       */
      let letArrowFunc = () => {
        console.log('letArrowFunc');
      }
      
      /**
       * constArrowFunc
       * @doc UI
       */
      const constArrowFunc = () => {
        console.log('constArrowFunc');
      }
      `;

        expect(
          buildDocumentationSource({
            source: sourceCode,
            tag: 'UI',
          }),
        ).toMatchSnapshot();
      });
    });
  });

  describe('by source entry', () => {
    it('should extract documentation source from entry dependencies', () => {
      expect(
        buildDocumentationSource({
          entry: require.resolve('./fixtures/src/index.ts'),
          tag: 'UI',
        }),
      ).toMatchSnapshot();

      expect(
        buildDocumentationSource({
          entry: require.resolve('./fixtures/src/index.ts'),
          tag: 'Events',
        }),
      ).toMatchSnapshot();
    });
  });

  describe('by source glob pattern', () => {
    it('should extract documentation source from files matched glob', () => {
      expect(
        buildDocumentationSource({
          pattern: `${path.join(__dirname, 'fixtures/src')}/**/*.ts`,
          tag: 'UI',
        }),
      ).toMatchSnapshot();

      expect(
        buildDocumentationSource({
          pattern: `${path.join(__dirname, 'fixtures/src')}/**/*.ts`,
          tag: 'utils',
        }),
      ).toMatchSnapshot();
    });
  });
});
