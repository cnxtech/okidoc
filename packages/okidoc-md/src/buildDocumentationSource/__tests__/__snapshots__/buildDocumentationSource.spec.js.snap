// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildDocumentationSource by source entry should extract documentation source from entry dependencies 1`] = `
"/** */
class API {
  /**
  * getter for \`isHidden\` value
  *
  * @example
  * console.log(component.isHidden) // false
  *
  * @returns \`true\` if hidden
  */
  get isHidden(): boolean {}

  /**
  * show
  *
  * @example
  * component.show()
  */
  show() {}

  /**
  * Hide
  *
  * @example
  * component.hide()
  */
  hide() {}

  /**
  * Get component node
  */
  getNode(): Node {}

  /**
  * Render component
  */
  render() {}

}"
`;

exports[`buildDocumentationSource by source entry should extract documentation source from entry dependencies 2`] = `
"/** */
class API {
  /**
  * add event listener
  */
  addEventListener(fn: Function) {}

  /**
  * remove event listener
  */
  removeEventListener(fn: Function) {}

}"
`;

exports[`buildDocumentationSource by source glob pattern should extract documentation source from files matched glob 1`] = `
"/** */
class API {
  /**
  * Get component node
  */
  getNode(): Node {}

  /**
  * Render component
  */
  render() {}

  /**
  * getter for \`isHidden\` value
  *
  * @example
  * console.log(component.isHidden) // false
  *
  * @returns \`true\` if hidden
  */
  get isHidden(): boolean {}

  /**
  * show
  *
  * @example
  * component.show()
  */
  show() {}

  /**
  * Hide
  *
  * @example
  * component.hide()
  */
  hide() {}

}"
`;

exports[`buildDocumentationSource by source glob pattern should extract documentation source from files matched glob 2`] = `
"/** */
class API {}

/**
* mountComponent
*
* 
*/
function mountComponent(component: Function, node: Node) {}

/**
* renderComponent
*
* 
*/
function renderComponent(component: Function) {}

/**
* unmountComponent
*
* 
*/
function unmountComponent(component: Function) {}"
`;

exports[`buildDocumentationSource for API class with @doc tag on class declaration should extract class members 1`] = `
"/** */
class API {
  /**
  * \`isHidden\` getter
  */
  get isHidden() {}

  /**
  * show
  */
  show() {}

  /**
  * hide
  */
  hide() {}

  /** */
  disable() {}

}"
`;

exports[`buildDocumentationSource for API class with @doc tag on class declaration should ignore class constructor and private members 1`] = `
"/** */
class API {
  /** */
  render() {}

}"
`;

exports[`buildDocumentationSource for API class with @doc tag on class members should extract getters and setters with tag 1`] = `
"/** */
class API {
  /**
  * \`width\` getter
  *
  * 
  */
  get width() {}

  /**
  * \`width\` setter
  *
  * 
  */
  set width(value) {}

}"
`;

exports[`buildDocumentationSource for API class with @doc tag on class members should extract methods with tag 1`] = `
"/** */
class API {
  /**
  * show
  *
  * 
  * @example
  * component.show()
  */
  show() {}

}"
`;

exports[`buildDocumentationSource for API class with @doc tag on class members should ignore methods with different tag 1`] = `
"/** */
class API {
  /**
  * show
  *
  * 
  */
  show() {}

}"
`;

exports[`buildDocumentationSource for API class with @doc tag on class members should ignore private methods with tag 1`] = `
"/** */
class API {}"
`;

exports[`buildDocumentationSource for API class with custom visitor should extract class members 1`] = `
"/** */
class API {
  /** */
  get isHidden() {}

  /**
  * show
  */
  show() {}

  /**
  * hide
  */
  hide() {}

}"
`;

exports[`buildDocumentationSource for functions with @doc tag on function declaration should extract functions with tag 1`] = `
"/** */
class API {}

/**
* showComponent
*
* 
*/
function showComponent() {}

/**
* hideComponent
*
* 
*/
function hideComponent() {}"
`;

exports[`buildDocumentationSource for functions with @doc tag on var/let/const declaration should extract functions with tag 1`] = `
"/** */
class API {}

/**
* varFunc
* 
*/
var varFunc = function () {
  console.log('varFunc');
};

/**
* varArrowFunc
* 
*/
var varArrowFunc = () => {
  console.log('varArrowFunc');
};

/**
* letArrowFunc
* 
*/
let letArrowFunc = () => {
  console.log('letArrowFunc');
};

/**
* constArrowFunc
* 
*/
const constArrowFunc = () => {
  console.log('constArrowFunc');
};"
`;
